// nagaris-frigg-backend/src/api-modules/twilio/definition.js
require('dotenv').config();
const { Api } = require('./api');
const config = require('./defaultConfig.json');

const Definition = {
    API: Api,
    getName: () => config.name,
    moduleName: config.name,
    modelName: 'Twilio',
    requiredAuthMethods: {
        getToken: async (api, params) => {
            // For basic auth, we just return the credentials passed in
            const accountSid = params.accountSid || params.data?.accountSid;
            const authToken = params.authToken || params.data?.authToken;

            if (!accountSid || !authToken) {
                throw new Error('Account SID and Auth Token are required');
            }

            return {
                accountSid,
                authToken
            };
        },
        getEntityDetails: async (api, callbackParams, tokenResponse, userId) => {
            const accountDetails = await api.getAccountDetails();
            return {
                identifiers: { externalId: accountDetails.sid, user: userId },
                details: { name: accountDetails.friendly_name || accountDetails.sid },
            };
        },
        apiPropertiesToPersist: {
            credential: ['accountSid', 'authToken'],
            entity: [],
        },
        getCredentialDetails: async (api, userId) => {
            const accountDetails = await api.getAccountDetails();
            return {
                identifiers: { externalId: accountDetails.sid, user: userId },
                details: {}
            };
        },
        testAuthRequest: async (api) => api.getAccountDetails(),
    },
    env: {
        // No OAuth env variables needed for basic auth
    }
};

module.exports = { Definition };
