// nagaris-frigg-backend/src/api-modules/simplefund360/definition.js
require('dotenv').config();
const { Api } = require('./Api');
const { get } = require('@friggframework/core');
const config = require('./defaultConfig.json');

const Definition = {
    API: Api,
    getName: () => config.name,
    moduleName: config.name,
    modelName: 'SimpleFund360',
    requiredAuthMethods: {
        getToken: async (api, params) => {
            const code = get(params.data, 'code');
            return api.getTokenFromCode(code);
        },
        getEntityDetails: async (api, callbackParams, tokenResponse, userId) => {
            // Get user/account details from BGL
            const userDetails = await api.getCurrentUser();

            return {
                identifiers: {
                    externalId: userDetails.id || userDetails.userId,
                    user: userId
                },
                details: {
                    name: userDetails.name || userDetails.email,
                },
            };
        },
        apiPropertiesToPersist: {
            credential: ['access_token', 'refresh_token'],
            entity: [],
        },
        getCredentialDetails: async (api, userId) => {
            const userDetails = await api.getCurrentUser();

            return {
                identifiers: {
                    externalId: userDetails.id || userDetails.userId,
                    user: userId
                },
                details: {}
            };
        },
        testAuthRequest: async (api) => api.getCurrentUser(),
    },
    env: {
        client_id: process.env.BGL_CLIENT_ID,
        client_secret: process.env.BGL_CLIENT_SECRET,
        scope: process.env.BGL_SCOPE || 'developer',
        redirect_uri: `${process.env.REDIRECT_URI}/bgl/callback`,
    }
};

module.exports = { Definition };
