// nagaris-frigg-backend/src/integrations/Cas360Integration.js
const { IntegrationBase } = require('@friggframework/core');
const cas360 = require('../api-modules/cas360');

class Cas360Integration extends IntegrationBase {
    static Definition = {
        name: 'cas360',
        version: '1.0.0',
        supportedVersions: ['1.0.0'],
        hasUserConfig: false,

        display: {
            label: 'BGL CAS360',
            description: 'Syncs Company and Trust clients with BGL CAS360 practice management software',
            category: 'Practice Management',
            detailsUrl: 'https://cas360.com.au',
            icon: 'https://www.bglcorp.com/wp-content/uploads/2021/03/favicon.png',
        },
        modules: {
            cas360: {
                definition: cas360.Definition,
            },
        },
        routes: [
            {
                path: '/auth',
                method: 'GET',
                event: 'AUTH_REQUEST',
            },
            {
                path: '/callback',
                method: 'GET',
                event: 'HANDLE_CALLBACK',
            },
            {
                path: '/companies/import',
                method: 'GET',
                event: 'IMPORT_COMPANIES',
            },
            {
                path: '/companies/sync',
                method: 'POST',
                event: 'SYNC_COMPANY',
            },
            {
                path: '/trusts/sync',
                method: 'POST',
                event: 'SYNC_TRUST',
            },
            {
                path: '/companies',
                method: 'GET',
                event: 'LIST_COMPANIES',
            },
            {
                path: '/trusts',
                method: 'GET',
                event: 'LIST_TRUSTS',
            },
        ],
    };

    constructor() {
        super();
        this.events = {
            AUTH_REQUEST: {
                handler: this.authRequest.bind(this),
            },
            HANDLE_CALLBACK: {
                handler: this.handleCallback.bind(this),
            },
            IMPORT_COMPANIES: {
                handler: this.importCompanies.bind(this),
            },
            SYNC_COMPANY: {
                handler: this.syncCompany.bind(this),
            },
            SYNC_TRUST: {
                handler: this.syncTrust.bind(this),
            },
            LIST_COMPANIES: {
                handler: this.listCompanies.bind(this),
            },
            LIST_TRUSTS: {
                handler: this.listTrusts.bind(this),
            },
        };
    }

    /**
     * Initiate OAuth flow
     */
    async authRequest({ req, res }) {
        const authUrl = this.cas360.api.authorizationUri;

        return res.json({
            url: authUrl,
            type: 'oauth2',
        });
    }

    /**
     * Handle OAuth callback
     */
    async handleCallback({ req, res }) {
        try {
            const { code } = req.query;

            if (!code) {
                return res.status(400).json({
                    error: 'Missing code',
                    message: 'Authorization code not found in callback',
                });
            }

            // Exchange code for token
            const tokenResponse = await this.cas360.api.getTokenFromCode(code);

            // Set credentials
            await this.cas360.api.setTokens(tokenResponse);

            // Test authentication
            try {
                await this.cas360.api.getCurrentUser();
            } catch (error) {
                return res.status(401).json({
                    error: 'Authentication failed',
                    message: 'Invalid token or unable to connect to CAS360',
                });
            }

            return res.json({
                success: true,
                message: 'Authentication successful',
                data: {
                    authenticated: true,
                },
            });
        } catch (error) {
            console.error('CAS360 callback error:', error);
            return res.status(500).json({
                error: 'Authentication failed',
                message: 'An error occurred during authentication',
            });
        }
    }

    /**
     * Import all companies from CAS360
     * Use case: Starting point to get full list of companies and match with Nagaris clients
     */
    async importCompanies({ req, res }) {
        try {
            const companies = await this.cas360.api.listCompanies();

            return res.json({
                success: true,
                message: 'Companies imported successfully',
                data: companies,
            });
        } catch (error) {
            console.error('Import companies error:', error);
            return res.status(500).json({
                error: 'Failed to import companies',
                message: error.message || 'An error occurred',
            });
        }
    }

    /**
     * Sync company to CAS360
     * Use case: When accountant creates/edits a Company client in Nagaris, push to CAS360
     */
    async syncCompany({ req, res }) {
        try {
            const { name, acn, abn, directors, address, additionalData } = req.body;

            if (!name) {
                return res.status(400).json({
                    error: 'Missing required fields',
                    message: 'Company name is required',
                });
            }

            const result = await this.cas360.api.syncCompany({
                name,
                acn,
                abn,
                directors,
                address,
                ...additionalData,
            });

            return res.json({
                success: true,
                message: 'Company synced to CAS360 successfully',
                data: result,
            });
        } catch (error) {
            console.error('Sync company error:', error);
            return res.status(500).json({
                error: 'Failed to sync company',
                message: error.message || 'An error occurred',
            });
        }
    }

    /**
     * Sync trust to CAS360
     * Use case: When accountant creates/edits a Trust client in Nagaris, push to CAS360
     */
    async syncTrust({ req, res }) {
        try {
            const { name, trustType, additionalData } = req.body;

            if (!name) {
                return res.status(400).json({
                    error: 'Missing required fields',
                    message: 'Trust name is required',
                });
            }

            const result = await this.cas360.api.syncTrust({
                name,
                trustType,
                ...additionalData,
            });

            return res.json({
                success: true,
                message: 'Trust synced to CAS360 successfully',
                data: result,
            });
        } catch (error) {
            console.error('Sync trust error:', error);
            return res.status(500).json({
                error: 'Failed to sync trust',
                message: error.message || 'An error occurred',
            });
        }
    }

    /**
     * List companies
     */
    async listCompanies({ req, res }) {
        try {
            const companies = await this.cas360.api.listCompanies(req.query);

            return res.json({
                success: true,
                data: companies,
            });
        } catch (error) {
            console.error('List companies error:', error);
            return res.status(500).json({
                error: 'Failed to list companies',
                message: error.message || 'An error occurred',
            });
        }
    }

    /**
     * List trusts
     */
    async listTrusts({ req, res }) {
        try {
            const trusts = await this.cas360.api.listTrusts(req.query);

            return res.json({
                success: true,
                data: trusts,
            });
        } catch (error) {
            console.error('List trusts error:', error);
            return res.status(500).json({
                error: 'Failed to list trusts',
                message: error.message || 'An error occurred',
            });
        }
    }
}

module.exports = Cas360Integration;
