// nagaris-frigg-backend/src/integrations/Simplefund360Integration.js
const { IntegrationBase } = require('@friggframework/core');
const simplefund360 = require('../api-modules/simplefund360');

/**
 * SimpleFund 360 Integration
 * SimpleFund
 */
class Simplefund360Integration extends IntegrationBase {
    static Definition = {
        name: 'simplefund360',
        version: '1.0.0',
        supportedVersions: ['1.0.0'],
        hasUserConfig: false,

        display: {
            label: 'SimpleFund 360',
            description: 'SimpleFund',
            category: 'Other',
            detailsUrl: '',
            icon: '',
        },
        modules: {
            simplefund360: {
                definition: simplefund360.Definition,
            },
        },
        routes: [
            {
                path: '/auth',
                method: 'GET',
                event: 'AUTH_REQUEST',
            },
        ],
    }

    constructor() {
        super();
        this.events = {
            AUTH_REQUEST: {
                handler: this.authRequest.bind(this),
            },
        };
    }

    // TODO: Add your integration methods here
    // Example:
    // async authRequest({ res }) {
    //     return res.json({ url: 'https://example.com/auth' });
    // }
}

module.exports = Simplefund360Integration;
