// nagaris-frigg-backend/src/integrations/TwilioIntegration.js
const { IntegrationBase } = require('@friggframework/core');
const twilio = require('../api-modules/twilio');

/**
 * Twilio Integration
 * Sending for all Nagaris users
 */
class TwilioIntegration extends IntegrationBase {
    static Definition = {
        name: 'twilio',
        version: '1.0.0',
        supportedVersions: ['1.0.0'],
        hasUserConfig: false,

        display: {
            label: 'Twilio',
            description: 'Sending for all Nagaris users',
            category: 'Communication',
            detailsUrl: '',
            icon: '',
        },
        modules: {
            // Add your API modules here
            twilio: {
                definition: twilio.Definition,
            },
            // Example:
            // myModule: {
            //     definition: myModule.Definition,
            // },
        },
        routes: [
            // Define your integration routes here
            // Example:
            // {
            //     path: '/auth',
            //     method: 'GET',
            //     event: 'AUTH_REQUEST',
            // },
        ],
    };

    constructor() {
        super();
        this.events = {
            // Define your event handlers here
            // Example:
            // AUTH_REQUEST: {
            //     handler: this.authRequest.bind(this),
            // },
        };
    }

    // TODO: Add your integration methods here
    // Example:
    // async authRequest({ res }) {
    //     return res.json({ url: 'https://example.com/auth' });
    // }
}

module.exports = TwilioIntegration;
